### PR TITLE
Pol 47 azure ahub utilization with manual entry

### PR DESCRIPTION
**Pull request by chandravurubindi**
### Description
Report when AHUB usage in Azure falls outside or inside the number of licenses specified by the user.

### Issues Resolved

This policy checks all instances in Azure to determine how many are using AHUB and report when that number falls outside or inside the specified license number.

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### Running Template link:
- Over Allocated: https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5d67b1d9de8dca00899710f7
- Under Allocated: https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5d67b2a5de8dca00899710f8

### Screenshots
- Over Allocated:
OverAllocated_Licenses
- Under Allocated:
UnderAllocated_License

### CODE REVIEW

File: compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+  validate $ds_vminstance_with_cores do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{data.summary_template}}"
+    detail_template <<-EOS
+# Azure AHUB Utilization with Manual Entry
+| Instance Name | Resource Group | Location | Instance Size | Image Publisher | Image Offer | Image Sku | Image Version |  Tags | Number Cores |
+| ------------- | -------------- | -------- | ------------- | --------------- | ----------- | --------- | ------------- |  ---- | ------------ |
+{{ range data.result -}}
+| {{.name}} | {{.resourceGroup}} | {{.location}} | {{.vmSize}} | {{.imagePublisher}} | {{.imageOffer}} | {{.imageSku}} | {{.imageVersion}} | {{.tags}} | {{.numberofcores}} |
+{{ end -}}
+#### Note: {{data.note}}
```
Comment (cdwilhelm) [inline compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt]: This isn't formatting correctly on the incident page

File: compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    end
+  end
+end
+
+datasource "ds_vminstance_with_cores" do
+  run_script $js_vminstances_with_cores, $ds_azure_vmsize, $ds_azure_instances, $param_exclusion_tag_key, $param_license_count
+end
+
+
+datasource "ds_results_size" do
```
Comment (cdwilhelm) [inline compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt]: I don't think this is necessary.  It appears you are only using this for the check.  in the check you can still use final_result.result.  You should be able use something like *check eq(size(val(data, "result")), 0) *
See example. https://github.com/flexera/policy_templates/blob/master/cost/downsize_instance/downsize_instance.pt#L332
You can remove ds_results_size and js_results_size.

File: compliance/azure/ahub_manual/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+- *Allowed AHUB licenses* - Number of AHUB licenses that are allowed to be run on Azure
+- *Exclusion Tag Key* - Azure VMs instance tag to ignore instance that are with AHUB enabled. Only supply the tag key. The policy assumes that the tag value is irrelevant.
+
+### Supported Clouds
+
+- Azure
+
+### Cost
+
+If the policy is used to remove AHUB on instances, those instances will be charged the on-demand rate for the given license.
```
Comment (cdwilhelm) [inline compliance/azure/ahub_manual/README.md]: This should change to "This policy does not incur any cloud costs."

File: compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+  label "Exclusion Tag Key"
+  description "Azure-native instance tag to ignore instances that are not using AHUB. Only supply the tag key. The policy assumes that the tag value is irrelevant."
+  type "string"
+  default "exclude_ahub"
+end
+
+parameter "param_license_count" do
+  type "string"
+  label "License Allowed Count"
+  description "Specify the number of licenses are allowed"
```
[the remark this answers is not in the document]
Comment (chandravurubindi) [reply compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt]: Made the required change

File: compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt (hunk 1 above)
Comment (chandravurubindi) [reply compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt]: Thanks, we made the required changes.

File: compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt (hunk 2 above)
Comment (chandravurubindi) [reply compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt]: Thank you Curt, we were not aware of this . Now I have taken care of this and tested the same.

File: compliance/azure/ahub_manual/README.md (hunk 3 above)
Comment (chandravurubindi) [reply compliance/azure/ahub_manual/README.md]: Updated Readme file. I hope the same needs to be changed in the requirement.

**Comment by chandravurubindi** [comment]
Hi Curt, 
Pls find the Running links after update
For Over allocated: https://governance.rightscale.com/org/1105/projects/60073/applied-policies/5d6e45892374e600a165842e
For under allocated: https://governance.rightscale.com/org/1105/projects/60073/applied-policies/5d6e44912374e600a165842d

**Pull request merged**